File: game/ui/HUD.ts
// 战斗 HUD：左上血条/等级/经验条，右上波次/击杀，中间波次提示，结算遮罩
// 挂载到 uiRoot（Director.ui），以屏幕中心为原点布局；数值变化时重绘/更新文本
import { Color, DrawNode, Label, Node, Vec2, View } from 'Dora';
import { ctx, type CombatStats } from 'game/core/GameContext';
import { clamp } from 'game/utils/MathUtils';

const FONT = 'sarasa-mono-sc-regular';

function createLabel(text: string, size: number, pos: Vec2.Type, anchor: Vec2.Type, parent: Node.Type): Label.Type | undefined {
	const label = Label(FONT, size);
	if (label === undefined) return undefined;
	label.text = text;
	label.anchor = anchor;
	label.position = pos;
	label.addTo(parent);
	return label;
}

function drawRect(draw: DrawNode.Type, x: number, y: number, w: number, h: number, color: Color.Type): void {
	draw.drawPolygon(
		[Vec2(x, y), Vec2(x + w, y), Vec2(x + w, y + h), Vec2(x, y + h)],
		color,
	);
}

export class HUD {
	private uiRoot: Node.Type;
	private root: Node.Type;
	// 血条/经验条
	private hpBg: DrawNode.Type;
	private hpFill: DrawNode.Type;
	private expBg: DrawNode.Type;
	private expFill: DrawNode.Type;
	private lvLabel: Label.Type | undefined;
	private hpLabel: Label.Type | undefined;
	private waveLabel: Label.Type | undefined;
	private killLabel: Label.Type | undefined;
	private modeLabel: Label.Type | undefined;
	private tipLabel: Label.Type | undefined;
	private tipTimer = 0;
	private tipActive = false;
	// 缓存上次渲染值，避免每帧重绘
	private lastHpRatio = -1;
	private lastExpRatio = -1;
	private lastLevel = -1;
	private lastWave = -1;
	private lastKills = -1;
	// 结算遮罩
	private resultRoot: Node.Type;
	private resultTitle: Label.Type | undefined;
	private resultStats: Label.Type | undefined;

	private readonly margin = 18;
	private readonly barW = 250;
	private readonly barH = 20;
	private readonly expH = 6;

	// root 为世界根（兼容参数），HUD 实际挂载到 uiRoot 保证 UI 层级
	constructor(root: Node.Type, uiRoot: Node.Type) {
		this.uiRoot = uiRoot;
		this.root = Node();
		this.root.addTo(uiRoot);

		const W = View.size.width;
		const H = View.size.height;
		const leftX = -W / 2 + this.margin;
		const topY = H / 2 - this.margin;
		const panels = DrawNode();
		panels.addTo(this.root);
		drawRect(panels, leftX - 10, topY - 76, this.barW + 142, 94, Color(8, 17, 20, 205));
		drawRect(panels, W / 2 - 225, topY - 103, 215, 121, Color(8, 17, 20, 205));
		panels.drawSegment(Vec2(leftX - 10, topY + 18), Vec2(leftX + this.barW + 132, topY + 18), 3, Color(205, 150, 55, 230));
		panels.drawSegment(Vec2(W / 2 - 225, topY + 18), Vec2(W / 2 - 10, topY + 18), 3, Color(70, 125, 121, 230));

		// 左上：等级
		this.lvLabel = createLabel('Lv.1', 30, Vec2(leftX, topY + 12), Vec2(0, 0.5), this.root);

		// 左上：血条（背景+填充）
		const hpY = topY - this.barH - 8;
		this.hpBg = DrawNode();
		this.hpBg.addTo(this.root);
		drawRect(this.hpBg, leftX, hpY, this.barW, this.barH, Color(35, 38, 46, 200));
		this.hpFill = DrawNode();
		this.hpFill.addTo(this.root);
		drawRect(this.hpFill, leftX, hpY, this.barW, this.barH, Color(226, 76, 86, 255));
		this.hpLabel = createLabel('100 / 100', 28, Vec2(leftX + this.barW + 12, hpY + this.barH / 2), Vec2(0, 0.5), this.root);

		// 左上：经验条（背景+填充）
		const expY = hpY - this.expH - 3;
		this.expBg = DrawNode();
		this.expBg.addTo(this.root);
		drawRect(this.expBg, leftX, expY, this.barW, this.expH, Color(35, 38, 46, 200));
		this.expFill = DrawNode();
		this.expFill.addTo(this.root);
		drawRect(this.expFill, leftX, expY, this.barW, this.expH, Color(140, 100, 230, 255));

		// 右上：波次/击杀
		this.waveLabel = createLabel('波次 1', 32, Vec2(W / 2 - this.margin, topY - 10), Vec2(1, 0.5), this.root);
		this.killLabel = createLabel('击杀 0', 28, Vec2(W / 2 - this.margin, topY - 46), Vec2(1, 0.5), this.root);
		this.modeLabel = createLabel('章节远征', 26, Vec2(W / 2 - this.margin, topY - 80), Vec2(1, 0.5), this.root);

		// 中间：波次提示（初始隐藏）
		this.tipLabel = createLabel('', 46, Vec2(0, H / 2 - 118), Vec2(0.5, 0.5), this.root);
		if (this.tipLabel !== undefined) {
			this.tipLabel.visible = false;
		}

		// 兼容旧调用的结算节点默认隐藏；实际结算由 PauseMenu 负责
		this.resultRoot = Node();
		this.resultRoot.addTo(this.uiRoot);
		const bg = DrawNode();
		bg.addTo(this.resultRoot);
		drawRect(bg, -W / 2, -H / 2, W, H, Color(8, 10, 18, 190));
		this.resultTitle = createLabel('', 40, Vec2(0, 70), Vec2(0.5, 0.5), this.resultRoot);
		this.resultStats = createLabel('', 18, Vec2(0, -20), Vec2(0.5, 0.5), this.resultRoot);
		this.resultRoot.visible = false;
	}

	// 每帧刷新 HUD 数值（仅变化时更新）
	update(dt: number): void {
		// 波次提示计时
		if (this.tipActive) {
			this.tipTimer -= dt;
			if (this.tipTimer <= 0) {
				this.tipActive = false;
				if (this.tipLabel !== undefined) this.tipLabel.visible = false;
			}
		}

		const player = ctx.player;
		if (player === undefined) return;
		if (this.hpLabel !== undefined) this.hpLabel.text = `${Math.ceil(player.hp)} / ${Math.ceil(player.maxHp)}`;

		// 血条
		const hpRatio = player.maxHp > 0 ? clamp(player.hp / player.maxHp, 0, 1) : 0;
		if (Math.abs(hpRatio - this.lastHpRatio) > 0.002) {
			this.lastHpRatio = hpRatio;
			const w = Math.max(0, this.barW * hpRatio);
			this.hpFill.clear();
			const hpY = View.size.height / 2 - this.margin - this.barH - 8;
			drawRect(this.hpFill, -View.size.width / 2 + this.margin, hpY, w, this.barH, Color(226, 76, 86, 255));
		}

		// 等级
		if (player.level !== this.lastLevel) {
			this.lastLevel = player.level;
			if (this.lvLabel !== undefined) this.lvLabel.text = `Lv.${player.level}`;
		}

		// 经验条
		const expRatio = player.expNeed > 0 ? clamp(player.exp / player.expNeed, 0, 1) : 0;
		if (Math.abs(expRatio - this.lastExpRatio) > 0.002) {
			this.lastExpRatio = expRatio;
			const w = Math.max(0, this.barW * expRatio);
			this.expFill.clear();
			const hpY = View.size.height / 2 - this.margin - this.barH - 8;
			const expY = hpY - this.expH - 3;
			drawRect(this.expFill, -View.size.width / 2 + this.margin, expY, w, this.expH, Color(140, 100, 230, 255));
		}

		// 右上统计
		const stats = ctx.stats;
		if (stats.wave !== this.lastWave) {
			this.lastWave = stats.wave;
			if (this.waveLabel !== undefined) this.waveLabel.text = `波次 ${stats.wave}`;
			if (this.modeLabel !== undefined) this.modeLabel.text = ctx.mode === 'endless' ? '无尽深潜' : ctx.mode === 'challenge' ? '高压试炼' : ctx.mode === 'daily' ? '每日秘境' : '章节远征';
		}
		if (stats.kills !== this.lastKills) {
			this.lastKills = stats.kills;
			if (this.killLabel !== undefined) this.killLabel.text = `击杀 ${stats.kills}`;
		}
	}

	// 波次提示（WaveManager 接线调用）
	showWaveTip(n: number): void {
		if (this.tipLabel === undefined) return;
		this.tipLabel.text = `第 ${n} 波来袭！`;
		this.tipLabel.visible = true;
		this.tipActive = true;
		this.tipTimer = 2.2;
	}

	// 游戏结束遮罩（结算面板细节由外壳域做）
	showGameOver(stats: CombatStats): void {
		this.showResult('游戏结束', stats);
	}

	// 胜利遮罩
	showVictory(stats: CombatStats): void {
		this.showResult('胜利！', stats);
	}

	private showResult(title: string, stats: CombatStats): void {
		if (this.resultTitle !== undefined) this.resultTitle.text = title;
		if (this.resultStats !== undefined) {
			this.resultStats.text =
				`第 ${stats.wave} 波    击杀 ${stats.kills}\n` +
				`生存 ${Math.floor(stats.timeAlive)} 秒    等级 ${stats.playerLevel}`;
		}
		this.resultRoot.visible = true;
	}

	// 隐藏结算遮罩
	hideResult(): void {
		this.resultRoot.visible = false;
	}

	// 重置 HUD（新对局时调用）
	reset(): void {
		this.lastHpRatio = -1;
		this.lastExpRatio = -1;
		this.lastLevel = -1;
		this.lastWave = -1;
		this.lastKills = -1;
		this.tipActive = false;
		if (this.tipLabel !== undefined) this.tipLabel.visible = false;
		this.hideResult();
	}
}

File: game/ui/HUD.lua
-- [ts]: HUD.ts
local ____lualib = require("lualib_bundle") -- 1
local __TS__Class = ____lualib.__TS__Class -- 1
local ____exports = {} -- 1
local ____Dora = require("Dora") -- 3
local Color = ____Dora.Color -- 3
local DrawNode = ____Dora.DrawNode -- 3
local Label = ____Dora.Label -- 3
local Node = ____Dora.Node -- 3
local Vec2 = ____Dora.Vec2 -- 3
local View = ____Dora.View -- 3
local ____GameContext = require("game.core.GameContext") -- 4
local ctx = ____GameContext.ctx -- 4
local ____MathUtils = require("game.utils.MathUtils") -- 5
local clamp = ____MathUtils.clamp -- 5
local FONT = "sarasa-mono-sc-regular" -- 7
local function createLabel(text, size, pos, anchor, parent) -- 9
	local label = Label(FONT, size) -- 10
	if label == nil then -- 10
		return nil -- 11
	end -- 11
	label.text = text -- 12
	label.anchor = anchor -- 13
	label.position = pos -- 14
	label:addTo(parent) -- 15
	return label -- 16
end -- 9
local function drawRect(draw, x, y, w, h, color) -- 19
	draw:drawPolygon( -- 20
		{ -- 21
			Vec2(x, y), -- 21
			Vec2(x + w, y), -- 21
			Vec2(x + w, y + h), -- 21
			Vec2(x, y + h) -- 21
		}, -- 21
		color -- 22
	) -- 22
end -- 19
____exports.HUD = __TS__Class() -- 26
local HUD = ____exports.HUD -- 26
HUD.name = "HUD" -- 26
function HUD.prototype.____constructor(self, root, uiRoot) -- 59
	self.tipTimer = 0 -- 40
	self.tipActive = false -- 41
	self.lastHpRatio = -1 -- 43
	self.lastExpRatio = -1 -- 44
	self.lastLevel = -1 -- 45
	self.lastWave = -1 -- 46
	self.lastKills = -1 -- 47
	self.margin = 18 -- 53
	self.barW = 250 -- 54
	self.barH = 20 -- 55
	self.expH = 6 -- 56
	self.uiRoot = uiRoot -- 60
	self.root = Node() -- 61
	self.root:addTo(uiRoot) -- 62
	local W = View.size.width -- 64
	local H = View.size.height -- 65
	local leftX = -W / 2 + self.margin -- 66
	local topY = H / 2 - self.margin -- 67
	local panels = DrawNode() -- 68
	panels:addTo(self.root) -- 69
	drawRect( -- 70
		panels, -- 70
		leftX - 10, -- 70
		topY - 76, -- 70
		self.barW + 142, -- 70
		94, -- 70
		Color(8, 17, 20, 205) -- 70
	) -- 70
	drawRect( -- 71
		panels, -- 71
		W / 2 - 225, -- 71
		topY - 103, -- 71
		215, -- 71
		121, -- 71
		Color(8, 17, 20, 205) -- 71
	) -- 71
	panels:drawSegment( -- 72
		Vec2(leftX - 10, topY + 18), -- 72
		Vec2(leftX + self.barW + 132, topY + 18), -- 72
		3, -- 72
		Color(205, 150, 55, 230) -- 72
	) -- 72
	panels:drawSegment( -- 73
		Vec2(W / 2 - 225, topY + 18), -- 73
		Vec2(W / 2 - 10, topY + 18), -- 73
		3, -- 73
		Color(70, 125, 121, 230) -- 73
	) -- 73
	self.lvLabel = createLabel( -- 76
		"Lv.1", -- 76
		30, -- 76
		Vec2(leftX, topY + 12), -- 76
		Vec2(0, 0.5), -- 76
		self.root -- 76
	) -- 76
	local hpY = topY - self.barH - 8 -- 79
	self.hpBg = DrawNode() -- 80
	self.hpBg:addTo(self.root) -- 81
	drawRect( -- 82
		self.hpBg, -- 82
		leftX, -- 82
		hpY, -- 82
		self.barW, -- 82
		self.barH, -- 82
		Color(35, 38, 46, 200) -- 82
	) -- 82
	self.hpFill = DrawNode() -- 83
	self.hpFill:addTo(self.root) -- 84
	drawRect( -- 85
		self.hpFill, -- 85
		leftX, -- 85
		hpY, -- 85
		self.barW, -- 85
		self.barH, -- 85
		Color(226, 76, 86, 255) -- 85
	) -- 85
	self.hpLabel = createLabel( -- 86
		"100 / 100", -- 86
		28, -- 86
		Vec2(leftX + self.barW + 12, hpY + self.barH / 2), -- 86
		Vec2(0, 0.5), -- 86
		self.root -- 86
	) -- 86
	local expY = hpY - self.expH - 3 -- 89
	self.expBg = DrawNode() -- 90
	self.expBg:addTo(self.root) -- 91
	drawRect( -- 92
		self.expBg, -- 92
		leftX, -- 92
		expY, -- 92
		self.barW, -- 92
		self.expH, -- 92
		Color(35, 38, 46, 200) -- 92
	) -- 92
	self.expFill = DrawNode() -- 93
	self.expFill:addTo(self.root) -- 94
	drawRect( -- 95
		self.expFill, -- 95
		leftX, -- 95
		expY, -- 95
		self.barW, -- 95
		self.expH, -- 95
		Color(140, 100, 230, 255) -- 95
	) -- 95
	self.waveLabel = createLabel( -- 98
		"波次 1", -- 98
		32, -- 98
		Vec2(W / 2 - self.margin, topY - 10), -- 98
		Vec2(1, 0.5), -- 98
		self.root -- 98
	) -- 98
	self.killLabel = createLabel( -- 99
		"击杀 0", -- 99
		28, -- 99
		Vec2(W / 2 - self.margin, topY - 46), -- 99
		Vec2(1, 0.5), -- 99
		self.root -- 99
	) -- 99
	self.modeLabel = createLabel( -- 100
		"章节远征", -- 100
		26, -- 100
		Vec2(W / 2 - self.margin, topY - 80), -- 100
		Vec2(1, 0.5), -- 100
		self.root -- 100
	) -- 100
	self.tipLabel = createLabel( -- 103
		"", -- 103
		46, -- 103
		Vec2(0, H / 2 - 118), -- 103
		Vec2(0.5, 0.5), -- 103
		self.root -- 103
	) -- 103
	if self.tipLabel ~= nil then -- 103
		self.tipLabel.visible = false -- 105
	end -- 105
	self.resultRoot = Node() -- 109
	self.resultRoot:addTo(self.uiRoot) -- 110
	local bg = DrawNode() -- 111
	bg:addTo(self.resultRoot) -- 112
	drawRect( -- 113
		bg, -- 113
		-W / 2, -- 113
		-H / 2, -- 113
		W, -- 113
		H, -- 113
		Color(8, 10, 18, 190) -- 113
	) -- 113
	self.resultTitle = createLabel( -- 114
		"", -- 114
		40, -- 114
		Vec2(0, 70), -- 114
		Vec2(0.5, 0.5), -- 114
		self.resultRoot -- 114
	) -- 114
	self.resultStats = createLabel( -- 115
		"", -- 115
		18, -- 115
		Vec2(0, -20), -- 115
		Vec2(0.5, 0.5), -- 115
		self.resultRoot -- 115
	) -- 115
	self.resultRoot.visible = false -- 116
end -- 59
function HUD.prototype.update(self, dt) -- 120
	if self.tipActive then -- 120
		self.tipTimer = self.tipTimer - dt -- 123
		if self.tipTimer <= 0 then -- 123
			self.tipActive = false -- 125
			if self.tipLabel ~= nil then -- 125
				self.tipLabel.visible = false -- 126
			end -- 126
		end -- 126
	end -- 126
	local player = ctx.player -- 130
	if player == nil then -- 130
		return -- 131
	end -- 131
	if self.hpLabel ~= nil then -- 131
		self.hpLabel.text = (tostring(math.ceil(player.hp)) .. " / ") .. tostring(math.ceil(player.maxHp)) -- 132
	end -- 132
	local hpRatio = player.maxHp > 0 and clamp(player.hp / player.maxHp, 0, 1) or 0 -- 135
	if math.abs(hpRatio - self.lastHpRatio) > 0.002 then -- 135
		self.lastHpRatio = hpRatio -- 137
		local w = math.max(0, self.barW * hpRatio) -- 138
		self.hpFill:clear() -- 139
		local hpY = View.size.height / 2 - self.margin - self.barH - 8 -- 140
		drawRect( -- 141
			self.hpFill, -- 141
			-View.size.width / 2 + self.margin, -- 141
			hpY, -- 141
			w, -- 141
			self.barH, -- 141
			Color(226, 76, 86, 255) -- 141
		) -- 141
	end -- 141
	if player.level ~= self.lastLevel then -- 141
		self.lastLevel = player.level -- 146
		if self.lvLabel ~= nil then -- 146
			self.lvLabel.text = "Lv." .. tostring(player.level) -- 147
		end -- 147
	end -- 147
	local expRatio = player.expNeed > 0 and clamp(player.exp / player.expNeed, 0, 1) or 0 -- 151
	if math.abs(expRatio - self.lastExpRatio) > 0.002 then -- 151
		self.lastExpRatio = expRatio -- 153
		local w = math.max(0, self.barW * expRatio) -- 154
		self.expFill:clear() -- 155
		local hpY = View.size.height / 2 - self.margin - self.barH - 8 -- 156
		local expY = hpY - self.expH - 3 -- 157
		drawRect( -- 158
			self.expFill, -- 158
			-View.size.width / 2 + self.margin, -- 158
			expY, -- 158
			w, -- 158
			self.expH, -- 158
			Color(140, 100, 230, 255) -- 158
		) -- 158
	end -- 158
	local stats = ctx.stats -- 162
	if stats.wave ~= self.lastWave then -- 162
		self.lastWave = stats.wave -- 164
		if self.waveLabel ~= nil then -- 164
			self.waveLabel.text = "波次 " .. tostring(stats.wave) -- 165
		end -- 165
		if self.modeLabel ~= nil then -- 165
			self.modeLabel.text = ctx.mode == "endless" and "无尽深潜" or (ctx.mode == "challenge" and "高压试炼" or (ctx.mode == "daily" and "每日秘境" or "章节远征")) -- 166
		end -- 166
	end -- 166
	if stats.kills ~= self.lastKills then -- 166
		self.lastKills = stats.kills -- 169
		if self.killLabel ~= nil then -- 169
			self.killLabel.text = "击杀 " .. tostring(stats.kills) -- 170
		end -- 170
	end -- 170
end -- 120
function HUD.prototype.showWaveTip(self, n) -- 175
	if self.tipLabel == nil then -- 175
		return -- 176
	end -- 176
	self.tipLabel.text = ("第 " .. tostring(n)) .. " 波来袭！" -- 177
	self.tipLabel.visible = true -- 178
	self.tipActive = true -- 179
	self.tipTimer = 2.2 -- 180
end -- 175
function HUD.prototype.showGameOver(self, stats) -- 184
	self:showResult("游戏结束", stats) -- 185
end -- 184
function HUD.prototype.showVictory(self, stats) -- 189
	self:showResult("胜利！", stats) -- 190
end -- 189
function HUD.prototype.showResult(self, title, stats) -- 193
	if self.resultTitle ~= nil then -- 193
		self.resultTitle.text = title -- 194
	end -- 194
	if self.resultStats ~= nil then -- 194
		self.resultStats.text = (((("第 " .. tostring(stats.wave)) .. " 波    击杀 ") .. tostring(stats.kills)) .. "\n") .. (("生存 " .. tostring(math.floor(stats.timeAlive))) .. " 秒    等级 ") .. tostring(stats.playerLevel) -- 196
	end -- 196
	self.resultRoot.visible = true -- 200
end -- 193
function HUD.prototype.hideResult(self) -- 204
	self.resultRoot.visible = false -- 205
end -- 204
function HUD.prototype.reset(self) -- 209
	self.lastHpRatio = -1 -- 210
	self.lastExpRatio = -1 -- 211
	self.lastLevel = -1 -- 212
	self.lastWave = -1 -- 213
	self.lastKills = -1 -- 214
	self.tipActive = false -- 215
	if self.tipLabel ~= nil then -- 215
		self.tipLabel.visible = false -- 216
	end -- 216
	self:hideResult() -- 217
end -- 209
return ____exports -- 209